### PR TITLE
[browser] Fix `HybridGlobalization`, SoftHyphen is not ignored when it should be

### LINKED ISSUE #95473: [browser] HybridGlobalization, SoftHypen is not ignored when it should be
See test data `StartsWith_StringComparison_TestData`. "\u00ADHello" and "Hello" is expected to be treated the same and as the consequence, "Hello".StartsWith("\u00ADHel") is expected to return true.
Should we add soft hypen to normalization step and ignore it for `StringComparison.CurrentCulture`? If we cannot do anything about it, add it to known platform differences in the docs.

The failing tests are using `mono_wasm_index_of`.

### PR DESCRIPTION
**Pull request by ilonatommy**
Fixes https://github.com/dotnet/runtime/issues/95473.

- `compare` / `endsWith` / `startsWith` ignored only:
`\u200B (Zero Width Space)`
`\u200C (Zero Width Non-Joiner)`
`\u200D (Zero Width Joiner)`
`\uFEFF (Zero Width No-Break Space)`
`\0 (Null character)`
With this PR they start ignoring `\u00AD` as well.

- `indexOf` did not ignore any of above because that would affect the index count. It turns out that `\u00AD` (SoftHyphen) does not matter in index count and it's correct to ignore it. Other characters ignored by comparison functions influence the index count.

- Moving the turkish-i issue from blocking full test to blocking specific test cases that fail

- Unblocking the test uncovered another, non related issue: `String.Replace` uses indexing function implementation with `matchLength` parameter. This implementation is not supported by `HybridGlobalization` on browser - `matchLength` should throw. This PR changes the expected value for tests that use `String.Replace` and adds comment in the docs.

- We were not throwing `PNSE` for IndexOf implementation with `matchLength` parameter even though we do not calculate it and similar methods, `IsPrefix` and `IsSuffix` throw `PNSE` by design. Now we start throwing correctly.
https://github.com/dotnet/runtime/blob/6931b3b729ec050f935f958122d3cc66bb46fdcf/docs/design/features/globalization-hybrid-mode.md?plain=1#L271

- Documentation got updated to reflect the actual state - we throw PNSE when methods with `matchLength` are used.

ToDo:
- iOS should also throw when `IndexOf` with `matchLength` is used. Make a follow-up PR with it.

### CODE REVIEW

File: src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs [hunk 1]
```diff
@@ -391,6 +391,7 @@ public static string GetDistroVersionString()
         public static bool IsHybridGlobalizationOnApplePlatform => m_isHybrid.Value && (IsMacCatalyst || IsiOS || IstvOS);
         public static bool IsNotHybridGlobalizationOnBrowser => !IsHybridGlobalizationOnBrowser;
         public static bool IsNotInvariantGlobalization => !IsInvariantGlobalization;
+        public static bool IsNotInvariantGlobalizationAndNotHybridGlobalizationOnBrowser => IsNotInvariantGlobalization && IsNotHybridGlobalizationOnBrowser;
```
Comment (matouskozak) [inline src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:394]: Is this needed? I don't see it used anywhere in this PR.

File: docs/design/features/globalization-hybrid-mode.md [hunk 2]
```diff
@@ -287,6 +286,15 @@ Affected public APIs:
 
 Web API does not expose locale-sensitive indexing function. There is a discussion on adding it: https://github.com/tc39/ecma402/issues/506. In the current state, as a workaround, locale-sensitive string segmenter combined with locale-sensitive comparison is used. This approach, beyond having the same compare option limitations as described under **String comparison**, has additional limitations connected with the workaround used. Information about additional limitations:
 
+- Methods that calculate `matchLength` always return throw PlatformNotSupported exception:
+
+[CompareInfo.IndexOf](https://learn.microsoft.com/dotnet/api/system.globalization.compareinfo.issuffix?view=system-globalization-compareinfo-issuffix(system-readonlyspan((system-char))-system-readonlyspan((system-char))-system-globalization-compareoptions-system-int32@))
```
Comment (matouskozak) [inline docs/design/features/globalization-hybrid-mode.md:291]: I think this should link to `IndexOf` and not to `IsSuffix` documentation.

**Review by matouskozak — COMMENTED** [review]
A few small nits but overall LGTM!

**Review by matouskozak — APPROVED** [review]

File: docs/design/features/globalization-hybrid-mode.md [hunk 3]
```diff
@@ -287,6 +286,15 @@ Affected public APIs:
...
 
+- Methods that calculate `matchLength` always return throw PlatformNotSupported exception:
+
+[CompareInfo.IndexOf](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.compareinfo.indexof?view=system-globalization-compareinfo-indexof(system-readonlyspan((system-char))-system-readonlyspan((system-char))-system-globalization-compareoptions-system-int32@))
+
+[CompareInfo.LastIndexOf](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.compareinfo.lastindexof?view=system-globalization-compareinfo-lastindexof(system-readonlyspan((system-char))-system-readonlyspan((system-char))-system-globalization-compareoptions-system-int32@))
+
+- String.Replace that uses `StringComparison` argument relies internally on IndexOf with `matchLength` argument. From this reason, it throws PlatformNotSupportedException:
+[String.Replace](https://learn.microsoft.com/en-us/dotnet/api/system.string.replace?view=system-string-replace(system-string-system-string-system-stringcomparison))
```
Comment (mkhamoyan) [inline docs/design/features/globalization-hybrid-mode.md:296]: We need to add this also under **String starts with / ends with** section yes?

File: docs/design/features/globalization-hybrid-mode.md [hunk 4]
```diff
@@ -463,6 +471,11 @@ Affected public APIs:
 - String.IndexOf
 - String.LastIndexOf
 
+Methods that calculate `matchLength` throw PlatformNotSupported exception:
+[CompareInfo.IndexOf](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.compareinfo.indexof?view=system-globalization-compareinfo-indexof(system-readonlyspan((system-char))-system-readonlyspan((system-char))-system-globalization-compareoptions-system-int32@))
+
+[CompareInfo.LastIndexOf](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.compareinfo.lastindexof?view=system-globalization-compareinfo-lastindexof(system-readonlyspan((system-char))-system-readonlyspan((system-char))-system-globalization-compareoptions-system-int32@))
+
```
Comment (mkhamoyan) [inline docs/design/features/globalization-hybrid-mode.md:478]: Same for Apple, let's add info about `PlatformNotSupported` exception for `IsPrefix/IsSuffix ` also.

**Review by mkhamoyan — APPROVED** [review]

File: docs/design/features/globalization-hybrid-mode.md (hunk 3 above)
Comment (ilonatommy) [reply docs/design/features/globalization-hybrid-mode.md:296]: We have it
browser:
https://github.com/dotnet/runtime/blob/7f315b171c47df5255f03def1a389369688bc92c/docs/design/features/globalization-hybrid-mode.md?plain=1#L271

ios:
https://github.com/dotnet/runtime/blob/7f315b171c47df5255f03def1a389369688bc92c/docs/design/features/globalization-hybrid-mode.md?plain=1#L449

Comment (mkhamoyan) [reply docs/design/features/globalization-hybrid-mode.md:296]: Thanks, missed that :)

**Pull request merged**